### PR TITLE
chore(trace): move context and provider modules

### PR DESCRIPTION
**Pull request by majorgreys**
Stacked after https://github.com/DataDog/dd-trace-py/pull/8234

## Checklist

- [x] Change(s) are motivated and described in the PR description
- [x] Testing strategy is described if automated tests are not included in the PR
- [x] Risks are described (performance impact, potential for breakage, maintainability)
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed or label `changelog/no-changelog` is set
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/))
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))
- [x] If this PR changes the public interface, I've notified `@DataDog/apm-tees`.
- [x] If change touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.

## Reviewer Checklist

- [x] Title is accurate
- [x] All changes are related to the pull request's stated goal
- [x] Description motivates each change
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- [x] Testing strategy adequately addresses listed risks
- [x] Change is maintainable (easy to change, telemetry, documentation)
- [x] Release note makes sense to a user of the library
- [x] Author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

**Review by romainkomorndatadog — APPROVED** [review]
LGTM (as in tests pass and that's about as much bandwidth as I have for this change at the moment).

File: ddtrace/internal/ci_visibility/context.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ContextTypeValue = t.Optional[t.Union[Context, Span]]
+
+
+_DD_CI_CONTEXTVAR: contextvars.ContextVar[ContextTypeValue] = contextvars.ContextVar(
+    "datadog_civisibility_contextvar",
+    default=None,
+)
+
+
+class CIContextProvider(DefaultContextProvider):
```
Comment (ZStriker19) [inline ddtrace/internal/ci_visibility/context.py:18]: This is a good separation, nice!

File: ddtrace/_trace/context.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
+import base64
```
Comment (emmettbutler) [inline ddtrace/_trace/context.py:1]: Have you considered combining the `_trace` and `tracing` directories? What's the distinction between them meant to be?

(It's a decision that was made elsewhere, but it's the first time I'm seeing it.)

**Review by emmettbutler — APPROVED** [review]

File: ddtrace/_trace/context.py (hunk 2 above)
Comment (majorgreys) [reply ddtrace/_trace/context.py:1]: I put it on my todo list, thanks!

**Review by P403n1x87 — APPROVED** [review]

**Pull request merged**